internal/kojiapi/server.go

```diff
@@ -0,0 +1,172 @@
+//go:generate go run github.com/deepmap/oapi-codegen/cmd/oapi-codegen --package=kojiapi --generate types,chi-server,client -o openapi.gen.go openapi.yml
+
+// Package kojiapi provides a REST API to build and push images to Koji
+package kojiapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/google/uuid"
+	"github.com/osbuild/osbuild-composer/internal/blueprint"
+	"github.com/osbuild/osbuild-composer/internal/distro"
+	"github.com/osbuild/osbuild-composer/internal/rpmmd"
+	"github.com/osbuild/osbuild-composer/internal/target"
+	"github.com/osbuild/osbuild-composer/internal/worker"
+)
+
+// Server represents the state of the koji Server
+type Server struct {
+	workers     *worker.Server
+	rpmMetadata rpmmd.RPMMD
+	distros     *distro.Registry
+}
+
+// NewServer creates a new koji server
+func NewServer(workers *worker.Server, rpmMetadata rpmmd.RPMMD, distros *distro.Registry) *Server {
+	server := &Server{
+		workers:     workers,
+		rpmMetadata: rpmMetadata,
+		distros:     distros,
+	}
+	return server
+}
+
+// Serve serves the koji API over the provided listener socket
+func (server *Server) Serve(listener net.Listener) error {
+	s := http.Server{Handler: Handler(server)}
+
+	err := s.Serve(listener)
+	if err != nil && err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// PostCompose handles a new /compose POST request
+func (server *Server) PostCompose(w http.ResponseWriter, r *http.Request) {
+	contentType := r.Header["Content-Type"]
+	if len(contentType) != 1 || contentType[0] != "application/json" {
+		http.Error(w, "Only 'application/json' content type is supported", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	var request ComposeRequest
+	err := json.NewDecoder(r.Body).Decode(&request)
+	if err != nil {
+		http.Error(w, "Could not parse JSON body", http.StatusBadRequest)
+		return
+	}
+
+	d := server.distros.GetDistro(request.Distribution)
+	if d == nil {
+		http.Error(w, fmt.Sprintf("Unsupported distribution: %s", request.Distribution), http.StatusBadRequest)
+		return
+	}
+
+	type imageRequest struct {
+		manifest distro.Manifest
+		filename string
+	}
+	imageRequests := make([]imageRequest, len(request.ImageRequests))
+
+	for i, ir := range request.ImageRequests {
+		arch, err := d.GetArch(ir.Architecture)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported architecture '%s' for distribution '%s'", ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		imageType, err := arch.GetImageType(ir.ImageType)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported image type '%s' for %s/%s", ir.ImageType, ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		repositories := make([]rpmmd.RepoConfig, len(ir.Repositories))
+		for j, repo := range ir.Repositories {
+			repositories[j].BaseURL = repo.Baseurl
+			repositories[j].GPGKey = repo.Gpgkey
+		}
+		packageSpecs, _ := imageType.Packages(blueprint.Blueprint{})
```

Comment:
Isn't `blueprint.Initialize()` required after constructing a new blueprint?

Reply:
Yes. It won't crash with the current implementation, but we should probably add it anyway.

Reply:
Fixed.